hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,233 @@
+        while True:
+            try:
+                for trading_pair in self._trading_pairs:
+                    snapshot_msg: OrderBookMessage = await self._order_book_snapshot(trading_pair)
+                    output.put_nowait(snapshot_msg)
+                    self.logger().debug(f"Saved order book snapshot for {trading_pair}")
+                delta = CONSTANTS.ONE_HOUR - time.time() % CONSTANTS.ONE_HOUR
+                await self._sleep(delta)
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().error(
+                    "Unexpected error occurred fetching orderbook snapshots. Retrying in 5 seconds...", exc_info=True
+                )
+                await self._sleep(5.0)
+
+    async def _parse_funding_info_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        if raw_message["type"] == "snapshot":
+            fields = raw_message.get("fields", [])
+            self._index_price_index = fields.index("indexRp")
+            self._mark_price_index = fields.index("markRp")
+            self._symbol_index = fields.index("symbol")
+            self._funding_rate_index = fields.index("fundingRateRr")
+
+        for data in raw_message.get("data", []):
+            trading_pair = await self._connector.trading_pair_associated_to_exchange_symbol(data[self._symbol_index])
+
+            if trading_pair not in self._trading_pairs:
+                continue
+            funding_info = FundingInfoUpdate(
+                trading_pair=trading_pair,
+                index_price=Decimal(data[self._index_price_index]),
+                mark_price=Decimal(data[self._mark_price_index]),
+                next_funding_utc_timestamp=self._next_funding_time(),
+                rate=Decimal(data[self._funding_rate_index]),
+            )
+
+            message_queue.put_nowait(funding_info)
+
+    async def _request_complete_funding_info(self, trading_pair: str):
+        ex_trading_pair = await self._connector.exchange_symbol_associated_to_pair(trading_pair=trading_pair)
+        data = await self._connector._api_get(
+            path_url=CONSTANTS.MARK_PRICE_URL, params={"symbol": ex_trading_pair}, is_auth_required=False
+        )
+        return data
+
+    def _next_funding_time(self) -> int:
+        """
+        Funding settlement occurs every 8 hours as mentioned in https://phemex.com/user-guides/funding-rate
+        """
+        return ((time.time() // 28800) + 1) * 28800
+
+    async def ping_loop(self, ws: WSAssistant):
+        while ws._connection.connected:
+            ping_request: WSJSONRequest = WSJSONRequest({"id": 0, "method": "server.ping", "params": []})
+            async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_MESSAGE_LIMIT_ID):
+                await ws.send(ping_request)
+            await self._sleep(5.0)
```

Comment:
The [docs](https://github.com/phemex/phemex-api-docs/blob/master/Public-Spot-API-en.md#session-management) say that the default ws ping-pong mechanism should be sufficient, which is already implemented when the ws is connected (i.e. `await ws.connect(ws_url=url, ping_timeout=CONSTANTS.WS_HEARTBEAT)`). 

However, if we do go for application-level ping-pong, then we should await and check if the pong was received. Also, and this one we can discuss since we don't have it as an accepted best practice, but I think we should avoid creating untracked async tasks (i.e. when we start the `ping_loop` on line 122 above). Bybit's API OB DS has an example of application-level ping-pong with pong tracking and keeping a reference to the loop so we can explicitly cancel it in the `_on_order_stream_interruption` method.

Reply:
@petioptrv please review the comment. Armada's standards and practices do not apply to the connectors we are sending to the open source version

Reply:
Just to clarify @vic-en, I think the change suggestion makes sense. I just think we should not discuss Armada related functionality in the open source PRs

Reply:
done